### PR TITLE
Provide UoA changes to upstream MyTardis

### PR DESCRIPTION
**Pull request by mike-ivs**
Hi all,

As promised (some time ago) here is a PR containing all the changes we made at UoA. We don't expect this PR to get merged as is as there are several combined features that should ideally be implemented into MT separately, but hopefully this PR will provide the groundwork to extract the relevant code without re-writing everything from scratch.

We'll try and document this PR to summarise the changes we've made, why we've made some of these changes, the extent and limitations of some new features, and potential pitfalls that might arise from them when rolling out to existing instances (i.e. fine grained ACLs). Please bear with us while we write this out, we'll try our best to be comprehensive so that you can extract the parts you want as easily as possible.

Cheers,
The Auckland Team (Mike, Chris, Noel, Warrick)

### **Outline of features**

- Project model
- Project overview page
- Search frontend
- Search metadata
- API changes (large!)
- ACLs deployed at Project/Dataset/Datafile levels
- Refactor of managers + decorators for ACLs
- New ACLs introduced: canSensitive + canDownload (previously merged with canRead)
- Numerous changes to reflect ACL requirements (counts/badges/downloads)
- Auto-indexing of newly created models (currently disabled)
- Group management expanded

### **Oustanding issues**

- Public functionality only at Exp level: effectively broken
- Some tests require public functionality to pass
- Tests required for new features (mostly ACL tests)
-  Auto-indexing of newly created models doesn't work well with GitHub tests yet (thus, disabled)
- Sharing and user/group/acl management needs developing out for proj/ds/dfs

### **Breakdown of features**

 **feature 1**

-

 **feature  2**

-

 **feature  3**

-

### CODE REVIEW

**Comment by manishkumr** [comment]
Thanks @mike-ivs and UoA team for submitting this PR. Looks like a lot of work being done here :clap: . This will help in implementing some of the much needed features in MyTardis. Thanks again :+1:

**Comment by manishkumr** [comment]
Hi @mike-ivs I had a chance to review this PR and looks like there are some good features developed here which we would like to merge upstream.
It would be good if we can separate this PR based on features developed.

Looking at the code changes, I believe the 

- ACLs at dataset/datafile level and changes related to introducing project models and related api changes could be part of one pull request.

- Also we might need to separate out migrations.


I am happy to work with you and get this merge. 

Let me know your thoughts on this.

Thanks again.

File: assets/js/apps/search/components/ResultSection.jsx [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        for (let key in resultSets) {
+            counts[key] = resultSets[key].length;
+        }
+    }
+
+        let selectedEntry = getSelectedEntry(resultSets, selectedResult, selectedType);
+
+        const currentResultSet = resultSets[selectedType],
+            currentCount = counts[selectedType];
+        return (
```
Comment (manishkumr) [inline assets/js/apps/search/components/ResultSection.jsx:224]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
    return (
```

File: assets/js/apps/search/components/filters/filters-section/FiltersSection.jsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
+import React, { Fragment } from 'react';
+import Tabs from "react-bootstrap/Tabs";
+import Tab from 'react-bootstrap/Tab';
+import { OBJECT_TYPE_STICKERS } from '../../TabStickers/TabSticker'
+import TypeSchemaList from '../type-schema-list/TypeSchemaList';
```
Comment (manishkumr) [inline assets/js/apps/search/components/filters/filters-section/FiltersSection.jsx:5]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
import TypeSchemaList from "../type-schema-list/TypeSchemaList";
```

File: assets/js/apps/search/components/filters/category-filter/CategoryFilter.stories.jsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import React, { useState } from 'react'
+import CategoryFilter  from './CategoryFilter';
```
Comment (manishkumr) [inline assets/js/apps/search/components/filters/category-filter/CategoryFilter.stories.jsx:2]: Codacy has a fix for the issue: [Multiple spaces found before 'from'.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
import CategoryFilter from './CategoryFilter';
```

File: assets/js/apps/project_detail_view/components/ExperimentList/ExperimentList.jsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      console.log('experiment by project fetch response.')
+      if (!response.ok) {
+        throw new Error("An error occurred fetching experiments.")
+      }
+      console.log("experiment fetch response was OK.")
+      response.json().then(experimentsData => {
+        setExperimentsData(experimentsData)
+      })
+    })
+  }, [])
```
Comment (manishkumr) [inline assets/js/apps/project_detail_view/components/ExperimentList/ExperimentList.jsx:27]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 2.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
    }, [])
```

File: assets/js/apps/search/components/PreviewCard/EntryPreviewCard.stories.jsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  title: 'EntryPreviewCard',
+  decorators: [story => <div style={{ 
+    border: '2px solid black'
+  }}>{story()}</div>],
+};
+
+const project = {
+  "type": "project",
+  "url": "idk/",
+  "counts": {
```
Comment (manishkumr) [inline assets/js/apps/search/components/PreviewCard/EntryPreviewCard.stories.jsx:16]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 2.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
    "counts": {
```

File: assets/js/apps/search/components/FilterSidebar.stories.jsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
```
Comment (manishkumr) [inline assets/js/apps/search/components/FilterSidebar.stories.jsx:1]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
import React from "react"
```

File: assets/js/apps/search/components/filters/range-filter/NumberRangeFilter.jsx [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        <Form className="num-range-filter" onSubmit={handleSubmit}>
+                    <Form.Group className="num-range-filter__field">
+                        <Form.Label>Min</Form.Label>
+                        <Form.Control  
+                            onChange={handleValueChange.bind(this,"min")} 
+                            value={localValue.min} 
+                            aria-label="Filter input for min value"
+                            placeholder={options.hintMin}
+                        >
+                        </Form.Control>
```
Comment (manishkumr) [inline assets/js/apps/search/components/filters/range-filter/NumberRangeFilter.jsx:105]: Codacy has a fix for the issue: [Expected indentation of 16 spaces but found 24.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
                </Form.Control>
```

File: assets/js/apps/search/components/filters/range-filter/NumberRangeFilter.jsx [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+                    </Form.Group>
+                    <Form.Group className="num-range-filter__field">
+                        <Form.Label>Max</Form.Label>
+                        <Form.Control 
+                            onChange={handleValueChange.bind(this,"max")} 
+                            value={localValue.max} 
+                            aria-label="Filter input for max value"
```
Comment (manishkumr) [inline assets/js/apps/search/components/filters/range-filter/NumberRangeFilter.jsx:112]: Codacy has a fix for the issue: [Expected indentation of 20 spaces but found 28.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
                    aria-label="Filter input for max value"
```

File: assets/js/apps/search/components/filters/filters-section/FiltersSection.jsx [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        value
+      }));
+      dispatch(runSearch());
+    });
+  };
+  const ApplicableFilter = mapTypeToFilter(attribute.data_type);
+  return (
+    <section>
+      <h3 className="h5">{attribute.full_name}</h3>
+      <ApplicableFilter id={typeId+"."+attributeId} value={attribute.value} onValueChange={setFilterValue} options={attribute.options} />
```
Comment (manishkumr) [inline assets/js/apps/search/components/filters/filters-section/FiltersSection.jsx:30]: Codacy has a fix for the issue: [Operator '+' must be spaced.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
      <ApplicableFilter id={typeId + "."+attributeId} value={attribute.value} onValueChange={setFilterValue} options={attribute.options} />
```

File: assets/js/apps/search/components/ResultList.stories.jsx [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    decorators: [story => <div style={{ padding: '3rem'}}>{story()}</div>],
+    excludeStories: /.*Data$/
+};
+
+export const allAccessResultData = {
+    id: "1",
+    type: "experiment",
+    title: "siNET",
+    url: "",
+    size:"35MB",
```
Comment (manishkumr) [inline assets/js/apps/search/components/ResultList.stories.jsx:17]: Codacy has a fix for the issue: [Missing space before value for key 'size'.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
    size: "35MB",
```

File: assets/js/apps/search/components/filters/date-filter/DateRangeFilter.jsx [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+                op: ">=",
+                content: localValue.start.toISOString()
+            });
+        }
+    }
+    if (!isNone(localValue.end)) {
+        if (typeof localValue.end == "object") {
+            submitValue.push({
+                op: "<=",
+                content:localValue.end.toISOString()
```
Comment (manishkumr) [inline assets/js/apps/search/components/filters/date-filter/DateRangeFilter.jsx:43]: Codacy has a fix for the issue: [Missing space before value for key 'content'.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
                content: localValue.end.toISOString()
```

File: assets/js/apps/search/components/filters/filterSlice.jsx [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+                const {kind ,target, content, op} = filter,
+                    value = {op, content};
+                let currValue, newValue;
+                switch (kind) {
+                    // TODO Refactor this function so that logic for 
+                    // adding filter values to existing can be reused
+                    // by both typeAttribute and schemaParameter.
+                    case "typeAttribute":
+                        // We may have added value for this field earlier in the iteration.
+                        currValue = typeAttrSelector(state,target[0],target[1]).value;
```
Comment (manishkumr) [inline assets/js/apps/search/components/filters/filterSlice.jsx:251]: Codacy has a fix for the issue: [A space is required after ','.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
                        currValue = typeAttrSelector(state, target[0],target[1]).value;
```

File: assets/js/apps/search/components/ResultSection.jsx [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+        selectedEntry = resultSets[selectedType].filter(result => result.id === selectedResult)[0];
+    }
+    return selectedEntry;
+}
+
+    export default function ResultSection() {
+        const selectedType = useSelector(state => state.search.selectedType),
+            selectedResult = useSelector(state => state.search.selectedResult),
+            dispatch = useDispatch(),
+            onSelectType = (type) => {
```
Comment (manishkumr) [inline assets/js/apps/search/components/ResultSection.jsx:265]: Codacy has a fix for the issue: [Expected indentation of 8 spaces but found 12.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
        onSelectType = (type) => {
```

File: assets/js/apps/search/components/filters/text-filter/TextFilter.jsx [hunk 14]
```diff
@@ -0,0 +1,67 @@
+import React, { useState, useEffect } from 'react';
+import Form from 'react-bootstrap/Form';
+import InputGroup from 'react-bootstrap/InputGroup';
+import FormControl from 'react-bootstrap/FormControl';
+import Button from 'react-bootstrap/Button';
+import PropTypes from 'prop-types';
+
+const TextFilter = ({value,options,onValueChange}) => {
```
Comment (manishkumr) [inline assets/js/apps/search/components/filters/text-filter/TextFilter.jsx:8]: Codacy has a fix for the issue: [A space is required after ','.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
const TextFilter = ({value, options,onValueChange}) => { 
```

File: assets/js/apps/project_detail_view/components/ProjectPage/ProjectPage.jsx [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    fetch(`/api/v1/project/?id=${props.projectId}`, {
+      method: 'get',
+      headers: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/json'
+      }
+    }).then(response => {
+      console.log(response)
+      setResponseStatus(response.status)
+      if (!response.ok) {
```
Comment (manishkumr) [inline assets/js/apps/project_detail_view/components/ProjectPage/ProjectPage.jsx:23]: Codacy has a fix for the issue: [Expected indentation of 12 spaces but found 6.](https://app.codacy.com/gh/mytardis/mytardis/pullRequest?prid=7933948)
```suggestion
            if (!response.ok) {
```

**Comment by mike-ivs** [comment]
Superseded by #3545

**Pull request closed without merging**